Extension documentation (#867)
qossmic/deptrac: docs/extending_deptrac.md

# Extending Deptrac

There are 2 main ways you can extend Deptrac:

- [Output formatters](#Output Formatters) - To change how the output of dependency analysis is formatted. This might be useful to integrate Deptrac into your CI/CD pipeline.

> - [Commands](#Commands) - Take advantage of the Deptrac framework to analyse different aspects of your dependency tree.

Author: I have found commands to be extremely useful extension points thus far. If you look at deptrac-awesome it is basically a Better Graphviz formatter and the rest are commands to get some useful information from the resolved dependencies.
For the extension point you have mentioned:
 - InputCollectorInterface - I do not see any reasons why you might want to do something other than what config allows right now.
 - DependencyEmitterInterface - This is useful, however, any additions here would be more than likely additions to core deptrac itself (like a new AST node that can play a role in dependency) rather than a custom extension
 - LayerResolverInterface, CollectorInterface, This I wanted to mainly cover in the "other extension points" section.

Apart from those, Deptrac also offers many other [extension points](#Other extension points) you can hook up into.

## Output Formatters

Creating an output formatter is very simple. Create a new class implementing the `Qossmic\Deptrac\OutputFormatter\OutputFormatterInterface` and register it in your `deptrac.yaml` file like this:

```yaml
services:
  - class: <Your FQCN>
    autowire: true
    tags:
      - output_formatter
```

And you are done. You can call your formatter by using the `-f` or `--formatter` CLI flag with the name you defined in `OutputFormatterInterface::getName()`.

## Commands

TODO

### Analyser Events

> - `ProcessEvent`

Author: Will do

# Extending Deptrac

> Deptrac defines its extension point by providing set of **contract** classes

Reviewer: ```suggestion
Deptrac defines its extension points by providing a set of **contract** classes
```

that you can depend on for your implementation. The classes can be found in
the `src/Contract` directory and are covered by

> the [backwards compatibility policy](bc_policy.md).

Reviewer: ```suggestion
the [backwards compatibility policy promise](bc_policy.md), meaning the will stay stable within major releases.
```

There are several ways you can extend Deptrac:

- [Output formatters](#Output Formatters) - To change how the output of
  dependency analysis is formatted. This might be useful to integrate Deptrac
  into your CI/CD pipeline.
- [Analyser Events](#Analyser Events) - Decide whether a dependency is
  Uncovered, Allowed, Violation or SkippedViolation.

> - [Layer collectors](#Layer collectors) - Add custom deptrac collector

Reviewer: ```suggestion
- [Layer collectors](#Layer collectors) - Add custom collectors for organizing your code into layers
```

> **Note**
> In examples where FQCN is not specified, the base
> namespace `Qossmic\Deptrac\Contract\` is omitted for readability.

## Output Formatters

> Creating an output formatter requires only to create a new class implementing

Reviewer: ```suggestion
Creating an output formatter requires creating a new class implementing
```

> the `OutputFormatter\OutputFormatterInterface` and

Reviewer: This might be hard to catch for people who don't know Deptrac's code well. I think this should be either `OutputFormatterInterface` to show, that it is not  in fact a FQCN or really use the FQCN instead.

Author: I mean, the note about this is literally 5 lines up.

Reviewer: Yes, but I think it's reasonable for people to see `OutputFormatter\OutputFormatterInterface` and assume this is already a FQCN. I think writing either the FQCN or just reducing it to the class name will make it easier to process for people. On the other hand, if you are extending deptrac we can assume you know or look at the classes in more detail anyway. Let's keep it for now and if someone is  confused or bothered we can change it later.

register it
in your `deptrac.yaml` file like this:

```yaml
services:
  - class: <Your FQCN>
    autowire: true
    tags:
      - output_formatter
```

And you are done. You can call your formatter by using the `-f` or `--formatter`
CLI flag with the name you defined

> in `OutputFormatter\OutputFormatterInterface::getName()`.

Reviewer: ```suggestion
in the `getName()`-method of your forrmatter.
```

> ## Analyser Events

Reviewer: It would be nice to have a tabular overview, which events Deptrac emits and what data it contains, but we can do this in a follow up.

Author: Right now it is only 2 events that are exposed in the contract. I don't think it currently warrants a table.

The deptrac analyser creates a `Analyser\ProcessEvent` for each dependency it
finds. You

> can listen for this event and yourself decide how to handle it depending on the

Reviewer: ```suggestion
can listen for this event and decide how to handle it depending on the
```

details of the dependency.

> This might be for example useful when you want to ignore a specific group of
> violations (in testing code for example). First you create a class to process

Reviewer: ```suggestion
This might be useful when you want to ignore a specific group of
violations in testing-related code for example. First you create a class to process
```

this event:

```php
class IgnoreDependenciesOnShouldNotHappenException
{
    public function __invoke(ProcessEvent $event): void
    {
        if ("Qossmic\Deptrac\Supportive\ShouldNotHappenException" === $event->dependentReference->getToken()->toString()) {
            $event->stopPropagation();
        }
    }
}
```

And then you register this class in your `deptrac.yaml` file:

```yaml
services:
  - class: IgnoreDependenciesOnShouldNotHappenException
    tags:
      - { name: kernel.event_listener, event: Qossmic\Deptrac\Contract\Analyser\ProcessEvent }
```

You can also en masse change the whole result set by instead listening to

> the `Analyser\PostProcessEvent`. This allows you to add Rules(Uncovered,

Reviewer: ```suggestion
the `Analyser\PostProcessEvent`. This allows you to add Rules (Uncovered,
```

Allowed, Violation, SkippedViolation),
Warnings and Errors to the analysis result. Do so by calling

> the `::replaceResult()` method.

Reviewer: ```suggestion
the `replaceResult()` method.
```

## Layer collectors

> Deptrac already comes with a good set of [collectors](collectors.md). If you

Reviewer: ```suggestion
Deptrac already comes with a comprehensive list of [collectors](collectors.md). If you
```

need something more specific, you can write your own collector
implementing `Layer\CollectorInterface` and register it in your `deptrac.yaml`
file:

```yaml
services:

>   - class: <Your FQCN>

Reviewer: ```suggestion
  - class: App\DeptracExtension\MyCustomLayerCollector
```
